Enter your name below in formate of "First_name Last_name"

In [1]:
NAME = "Janak Patil"

---

Project - 1

- To download data, go to this link and click download from data tab. You need to have Kaggle account for this.
https://www.kaggle.com/jboysen/us-perm-visas

Instructions for deliverables:
- Do not change anything in any other cell except in cells with comment "# YOUR CODE HERE" or "YOUR ANSWER HERE". If you need to play around with data, play around with it in another notebook. But donot submit that extra work as part of this notebook.
- Do not rename this notebook. Submit file as project1.ipynb only.
- Submit only one notebook. Before submitting, make sure everything runs as expected. To check that, restart the kernel (in the menubar, select Kernel > Restart) and then run all cells (in the menubar, select Cell > Run All).

Instructions for coding:
- Put the dataset in same folder as this file.
- Do not change the name of the dataset. Keep it as "us_perm_visas.csv" only.
- Use groupby in almost all questions.
- Use column 'case_number' for counting purposes. This means if this column has missing value, we don't count that row even if other columns have values in them.
- If same data is avaialble in multiple columns, use column with less missing values.
- Ignore missing data.
- Your output should show exactly what has been asked. Do not show whole big table with all the information. e.g. if it has been asked to list top 3 countries, your out put should have only 3 countries and not list of all countries.

Hint : All questions can be answered by using only one line of code. So try to use one line of code only to get answer.

In [2]:
import warnings
warnings.simplefilter('ignore')
import pandas as pd
data = pd.read_csv("us_perm_visas.csv")

In [3]:
data.head()

add_these_pw_job_title_9089 agent_city agent_firm_name agent_state  \
0                         NaN        NaN             NaN         NaN   
1                         NaN        NaN             NaN         NaN   
2                         NaN        NaN             NaN         NaN   
3                         NaN        NaN             NaN         NaN   
4                         NaN        NaN             NaN         NaN   

  application_type        case_no case_number case_received_date case_status  \
0             PERM  A-07323-97014         NaN                NaN   Certified   
1             PERM  A-07332-99439         NaN                NaN      Denied   
2             PERM  A-07333-99643         NaN                NaN   Certified   
3             PERM  A-07339-01930         NaN                NaN   Certified   
4             PERM  A-07345-03565         NaN                NaN   Certified   

  class_of_admission              ...               ri_pvt_employment_firm_to  \
0                J-1              ...                                     NaN   
1                B-2              ...                                     NaN   
2               H-1B              ...                                     NaN   
3                B-2              ...                                     NaN   
4                L-1              ...                                     NaN   

  ri_us_workers_considered schd_a_sheepherder     us_economic_sector  \
0                      NaN                NaN                     IT   
1                      NaN                NaN  Other Economic Sector   
2                      NaN                NaN              Aerospace   
3                      NaN                NaN  Other Economic Sector   
4                      NaN                NaN           Advanced Mfg   

  wage_offer_from_9089 wage_offer_to_9089 wage_offer_unit_of_pay_9089  \
0                75629                NaN                          yr   
1                37024                NaN                          yr   
2                47923                NaN                          yr   
3                10.97                NaN                          hr   
4               100000                NaN                          yr   

  wage_offered_from_9089 wage_offered_to_9089  wage_offered_unit_of_pay_9089  
0                    NaN                  NaN                            NaN  
1                    NaN                  NaN                            NaN  
2                    NaN                  NaN                            NaN  
3                    NaN                  NaN                            NaN  
4                    NaN                  NaN                            NaN  

[5 rows x 154 columns]

In [4]:
data.shape

(374362, 154)

In [5]:
for i in data.columns:
    print(i)

add_these_pw_job_title_9089
agent_city
agent_firm_name
agent_state
application_type
case_no
case_number
case_received_date
case_status
class_of_admission
country_of_citizenship
country_of_citzenship
decision_date
employer_address_1
employer_address_2
employer_city
employer_country
employer_decl_info_title
employer_name
employer_num_employees
employer_phone
employer_phone_ext
employer_postal_code
employer_state
employer_yr_estab
foreign_worker_info_alt_edu_experience
foreign_worker_info_birth_country
foreign_worker_info_city
foreign_worker_info_education
foreign_worker_info_education_other
foreign_worker_info_inst
foreign_worker_info_major
foreign_worker_info_postal_code
foreign_worker_info_rel_occup_exp
foreign_worker_info_req_experience
foreign_worker_info_state
foreign_worker_info_training_comp
foreign_worker_ownership_interest
foreign_worker_yr_rel_edu_completed
fw_info_alt_edu_experience
fw_info_birth_country
fw_info_education_other
fw_info_postal_code
fw_info_rel_occup_exp
fw_info

Q1. What is the count of applications submitted by each country?

In [6]:
data.groupby('country_of_citizenship')['case_number'].count()

country_of_citizenship
AFGHANISTAN                   16
ALBANIA                       93
ALGERIA                       29
ANDORRA                        1
ANGOLA                        13
ANTIGUA AND BARBUDA            9
ARGENTINA                    760
ARMENIA                      107
ARUBA                          0
AUSTRALIA                   1092
AUSTRIA                      143
AZERBAIJAN                    46
BAHAMAS                       68
BAHRAIN                        7
BANGLADESH                   700
BARBADOS                      29
BELARUS                      329
BELGIUM                      150
BELIZE                        25
BENIN                         18
BERMUDA                        6
BHUTAN                         6
BOLIVIA                      113
BOSNIA AND HERZEGOVINA        36
BOTSWANA                      13
BRAZIL                      1900
BRITISH VIRGIN ISLANDS         2
BRUNEI                         3
BULGARIA                     377
BURKINA FASO        

Q2. List top 3 countries with highest number of applications and also show number of applications by them.

In [7]:
data.groupby('country_of_citizenship')['case_number'].count().sort_values(ascending=False).head(3)

country_of_citizenship
INDIA          130707
CHINA           20177
SOUTH KOREA     18209
Name: case_number, dtype: int64

Q3. How many applications were denied?

In [9]:
data[data['case_status'] == 'Denied']['case_number'].count()

12781

Q4. How many applicants holding H-1B visas were certified for Permenant Residency? (Consider all forms used for writing "H1-B" in data. Consider only H1-B, not H1-B1)

In [10]:
data[(data.case_status == 'Certified') & ((data.class_of_admission == 'H-1B')|(data.class_of_admission == 'H1B'))].groupby('case_status')['case_number'].count()

case_status
Certified    88890
Name: case_number, dtype: int64

Q5. Sort columns in descending order according to number of missing values in them.

In [11]:
data.isna().sum().sort_values(ascending=False)

orig_file_date                         374081
orig_case_no                           374068
recr_info_job_fair_to                  372569
recr_info_job_fair_from                372564
recr_info_on_campus_recr_to            372497
recr_info_on_campus_recr_from          372490
ji_live_in_dom_svc_contract            371964
foreign_worker_info_education_other    370422
job_info_training_field                369883
job_info_training_num_months           369583
ri_coll_teach_select_date              368632
ri_coll_tch_basic_process              367349
recr_info_coll_teach_comp_proc         367341
job_info_education_other               366593
job_info_alt_combo_ed_other            365148
recr_info_pro_org_advert_to            360292
recr_info_pro_org_advert_from          360286
pw_source_name_other_9089              358603
ri_pvt_employment_firm_to              357922
ri_pvt_employment_firm_from            357920
ri_us_workers_considered               357129
naics_2007_us_title               

Q6. List top 10 cities from which employers have submitted the most applications. How many applicarions were submitted by them?

In [12]:
data.groupby('employer_city')['case_number'].count().sort_values(ascending=False).head(10)

employer_city
COLLEGE STATION    11632
NEW YORK            9844
SANTA CLARA         6966
SAN JOSE            6073
MOUNTAIN VIEW       6028
REDMOND             4901
SAN FRANCISCO       4125
SUNNYVALE           4075
PLANO               3844
CHICAGO             3784
Name: case_number, dtype: int64

Q7. Which are the top 10 employers who have submitted the highest number of applications? How many?

In [13]:
data.groupby('employer_name')['case_number'].count().sort_values(ascending=False).head(10)

employer_name
COGNIZANT TECHNOLOGY SOLUTIONS US CORPORATION    11587
MICROSOFT CORPORATION                             4625
INTEL CORPORATION                                 4127
GOOGLE INC.                                       3771
AMAZON CORPORATE LLC                              2716
CISCO SYSTEMS, INC.                               2097
INFOSYS LTD.                                      1795
APPLE INC.                                        1730
ORACLE AMERICA, INC.                              1552
FACEBOOK, INC.                                    1268
Name: case_number, dtype: int64

Q8. What is the average time of processing the applications?

In [14]:
Case_Recieved = (pd.to_datetime(data['case_received_date']))
Decision_Date = (pd.to_datetime(data['decision_date']))
Decision_Time = Decision_Date-Case_Recieved
Decision_Time.mean()

Timedelta('208 days 01:59:13.654466')

Q9. What is count of applications for each case status(e.g. Certified, Denied etc) for applicants with "Doctorate" degree?

In [15]:
data[data['foreign_worker_info_education'] == 'Doctorate'].groupby('case_status')['case_number'].count()

case_status
Certified            5965
Certified-Expired    5449
Denied                345
Withdrawn             484
Name: case_number, dtype: int64

Q.10 What are the top 3 economic sectors that has maximum applicants? How many applications are submitted from these sectors? (Use case_no column here, instead of case_number)

In [16]:
data.groupby('us_economic_sector')['case_no'].count().sort_values(ascending=False).head(3)

us_economic_sector
IT                       52617
Advanced Mfg             18749
Other Economic Sector    15798
Name: case_no, dtype: int64

Q.11 What is the mean wage of all the certified visa applications?

In [17]:
data[data['case_status']=='Certified'].groupby('case_status')['wage_offered_from_9089'].mean()

case_status
Certified    98359.286042
Name: wage_offered_from_9089, dtype: float64

Q.12 Show one instresting finding of your own which is not covered above. Show the code and write your finding as comment.

In [18]:
data['pw_amount_9089'] = pd.to_numeric(data['pw_amount_9089'],errors = 'coerce')
data.groupby('foreign_worker_info_education')['pw_amount_9089'].mean().sort_values(ascending=False)
## The finding below showcases what kind of degree would fetch what amount of average salary 

foreign_worker_info_education
Other          95196.657619
Master's       93038.799610
Bachelor's     91648.523886
Doctorate      77610.529559
Associate's    46694.148947
None           29220.897894
High School    28992.438391
Name: pw_amount_9089, dtype: float64

Q.13 Write a report on your analysis of this dataset. (Around 200 words, but quality of your report will be taken into account more than count of words)

•	The dataset consists of 374362 rows with 154 variables. Most citizens applying for permanent citizenship are from India, China, South Korea respectively.
•	Out of the 374362 people who applied, 12781 were denied permanent Visa
•	88890 people with H-1B visas were certified for permanent residency
•	College Station (11632) has the highest number of employers who have submitted an application for permanent visa, followed by NewYork(9844) and Santa Clara.
•	Top employers who have submitted application for permanent Visa are Cognizant Technology solutions US Corporation (11587) followed by Microsoft (4625), Intel (4127) and Google Inc (3771). It is evident by this that most companies who have applied are Information Technology related.
•	An average of 208 days is required to process a Visa application.
•	The ratio of certified Visa application for a person with “Doctorate” is much higher than for a “Master’s” or other degree.
•	IT sector (52617) has the highest number of people to have applied for Permanent Visa application. Advanced Mfg. (18749) follows this but only with a huge gap of 33868. 
•	The mean wage for people who have certified application is (98359$).

	The anomaly I found out from the individual interesting finding is that the average salary of “Doctorate” degree is less than the average salary for “Master’s” and “Bachelor’s”
	Other interesting finding not shown in the code I found out was the chance of getting a permanent Visa certified is higher for a person with a Doctorate degree.
	For every 1 visa application rejected 9 are accepted. The acceptance ratio is 9:1
	One final interesting finding is that the Cognizant Technology Solution US Corporation has the maximum number of Indian people with Certified Visas. This is probably because of its Indian roots.




In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 374362 entries, 0 to 374361
Columns: 154 entries, add_these_pw_job_title_9089 to wage_offered_unit_of_pay_9089
dtypes: float64(10), object(144)
memory usage: 439.8+ MB


In [4]:
data.describe()

employer_num_employees  employer_yr_estab  \
count            2.390130e+05      238958.000000   
mean             2.309387e+04        1984.208267   
std              5.453129e+05          38.470555   
min              0.000000e+00           0.000000   
25%              7.300000e+01        1980.000000   
50%              1.200000e+03        1996.000000   
75%              1.613400e+04        2004.000000   
max              2.635506e+08        2016.000000   

       foreign_worker_yr_rel_edu_completed  fw_info_yr_rel_edu_completed  \
count                         83364.000000                 137198.000000   
mean                           2001.345761                   2002.633603   
std                              91.444715                     89.674423   
min                               0.000000                      0.000000   
25%                            2003.000000                   2004.000000   
50%                            2006.000000                   2007.000000   
75%                            2010.000000                   2011.000000   
max                            2103.000000                   2207.000000   

       job_info_alt_cmb_ed_oth_yrs  job_info_alt_occ_num_months  \
count                 70220.000000                147200.000000   
mean                      4.593876                    36.483118   
std                       4.838582                    23.391668   
min                       0.000000                     0.000000   
25%                       3.000000                    24.000000   
50%                       5.000000                    24.000000   
75%                       5.000000                    60.000000   
max                      96.000000                   240.000000   

       job_info_experience_num_months  job_info_training_num_months  \
count                   133188.000000                   4779.000000   
mean                        33.103560                     32.685081   
std                         22.733393                     20.593288   
min                          0.000000                      0.000000   
25%                         12.000000                     12.000000   
50%                         24.000000                     36.000000   
75%                         60.000000                     36.000000   
max                        244.000000                    240.000000   

       wage_offered_from_9089  wage_offered_to_9089  
count            1.146470e+05          3.101500e+04  
mean             9.588846e+04          1.232273e+05  
std              1.463822e+05          1.080289e+05  
min              6.700000e+00          0.000000e+00  
25%              7.453400e+04          9.700100e+04  
50%              9.064400e+04          1.206000e+05  
75%              1.100000e+05          1.448419e+05  
max              1.629060e+07          1.328500e+07

In [7]:
data.describe(include=['O'])

add_these_pw_job_title_9089 agent_city  \
count                               42194     211454   
unique                               4276       1838   
top     Software Developers, Applications   New York   
freq                                10042      19359   

                                agent_firm_name agent_state application_type  \
count                                    206997      208239           135269   
unique                                    10135         108                3   
top     Fragomen, Del Rey, Bernsen & Loewy, LLP          CA           ONLINE   
freq                                      29569       35299           112564   

              case_no    case_number case_received_date case_status  \
count          135269         239093             239091      374362   
unique         134990         238418               2180           4   
top     A-13030-36063  A-14160-76066         2014-06-30   Certified   
freq                2              3                767      181933   

       class_of_admission              ...               \
count              351517              ...                
unique                 57              ...                
top                  H-1B              ...                
freq               283018              ...                

       ri_posted_notice_at_worksite ri_pvt_employment_firm_from  \
count                        239002                       16442   
unique                            3                        1006   
top                               Y                  2013-07-23   
freq                         235182                        2478   

       ri_pvt_employment_firm_to ri_us_workers_considered schd_a_sheepherder  \
count                      16440                    17233             239074   
unique                      1173                        3                  1   
top                   2013-07-30                        A                  N   
freq                        2478                     8892             239074   

       us_economic_sector wage_offer_from_9089 wage_offer_to_9089  \
count              129068               259591            73634.0   
unique                 17                38945            14053.0   
top                    IT            72,467.00           160000.0   
freq                52617                 2775             1519.0   

       wage_offer_unit_of_pay_9089 wage_offered_unit_of_pay_9089  
count                       258446                         70699  
unique                          10                             5  
top                           Year                          Year  
freq                        230591                         68950  

[4 rows x 144 columns]